### ISSUE TITLE
Fatal error while compiling Tensorflow with CUDA 9.1

### ISSUE THREAD

**Issue by icyblade** [tensorflow/tensorflow#15389]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.04.3 LTS (GNU/Linux 4.4.0-104-generic x86_64)
- **TensorFlow installed from (source or binary)**:
Source
- **TensorFlow version (use command below)**:
unknown 1.4.0 (Source code is cloned from 798fa36d11119e6fdc13b90a14abfe1805e7de90)
- **Python version**: 
3.6.3
- **Bazel version (if compiling from source)**:
0.8.1
- **GCC/Compiler version (if compiling from source)**:
gcc version 5.4.0 20160609
- **CUDA/cuDNN version**:
CUDA 9.1
cuDNN 7.0.5
- **GPU model and memory**:
2 * Tesla V100-PCIE-16GB
- **Exact command to reproduce**:
See description below.

### Describe the problem
While trying to compile the latest TensorFlow(cloned from 798fa36d11119e6fdc13b90a14abfe1805e7de90), such error will be raised:
```
ERROR: /home/ubuntu/tensorflow/tensorflow/contrib/seq2seq/BUILD:64:1: error while parsing .d file: /home/ubuntu/.cache/bazel/_bazel_ubuntu/ad1e09741bb4109fbc70ef8216b59ee2/execroot/org_tensorflow/bazel-out/k8-py3-opt/bin/tensorflow/contrib/seq2seq/_objs/python/ops/_beam_search_ops_gpu/tensorflow/contrib/seq2seq/kernels/beam_search_ops_gpu.cu.pic.d (No such file or directory)
In file included from external/eigen_archive/unsupported/Eigen/CXX11/Tensor:14:0,
                 from ./third_party/eigen3/unsupported/Eigen/CXX11/Tensor:1,
                 from ./tensorflow/contrib/seq2seq/kernels/beam_search_ops.h:19,
                 from tensorflow/contrib/seq2seq/kernels/beam_search_ops_gpu.cu.cc:20:
external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/Core:59:34: fatal error: math_functions.hpp: No such file or directory
```

It turns out that in CUDA 9.1, `math_functions.hpp` is located at `cuda/include/crt/math_functions.hpp`, rather than `cuda/include/math_functions.hpp` (CUDA 9.0 does), which leads to this error.
`ln -s /usr/local/cuda/include/crt/math_functions.hpp /usr/local/cuda/include/math_functions.hpp` will fix this problem and complete the compiling process.

#### Reference
https://stackoverflow.com/a/47807106/2666624

### Source code / logs
Traceback is available above.

**Comment by icyblade** [comment]
I'll try to open a PR resolving this problem, after digging some source code.

**Comment by icyblade** [comment]
I can confirm it's a bug of [Eigen](http://eigen.tuxfamily.org/).
Such error will be raised while compiling Eigen ([commit](https://bitbucket.org/eigen/eigen/src/5c43a259f0b51963a00dd85e941a1fd044ac1cf4)) with `cmake .. -DEIGEN_TEST_CUDA=ON && make check -j4`:
```
In file included from /home/ubuntu/eigen-eigen-5c43a259f0b5/Eigen/QR:11:0,
                 from /home/ubuntu/eigen-eigen-5c43a259f0b5/test/main.h:272,
                 from /home/ubuntu/eigen-eigen-5c43a259f0b5/test/cuda_basic.cu:23:
/home/ubuntu/eigen-eigen-5c43a259f0b5/Eigen/Core:59:34: fatal error: math_functions.hpp: No such file or directory
```

**Comment by AlessioTonioni** [comment]
https://stackoverflow.com/questions/43113508/math-functions-hpp-not-found-when-using-cuda-with-eigen --> second answer seems to be related

**Comment by icyblade** [comment]
@AlessioTonioni Yes, and I've already referenced this answer above.

**Comment by angersson** [comment]
Can you provide an exact command to help replicate this?

**Comment by sclarkson** [comment]
Also discussed in https://github.com/tensorflow/tensorflow/issues/15140

**Comment by icyblade** [comment]
@angersson

**Comment by hadaev8** [comment]
Same on cuda 9.0

**Comment by snnn** [comment]
It's a known issue. I can reproduce it under Fedora 25 and CUDA 9.1.

**Comment by reedwm** [comment]
/CC @nluehr can you fix this in Eigen?

**Comment by nluehr** [comment]
Issued [PR](https://bitbucket.org/eigen/eigen/pull-requests/353) to Eigen which should resolve this issue.

**Comment by JamesTheZ** [comment]
Make a softlink from /usr/local/cuda-9.1/include/crt/math_functions.hpp to /usr/local/cuda-9.1/include/math_functions.hpp can solve the problem for me.

**Comment by icyblade** [comment]
@JamesTheZ Which has already mentioned above, twice.

**Comment by arunmandal53** [comment]
Visit http://www.python36.com/install-tensorflow141-gpu/ for step by step instructions to build tensorflow from source for cuda 9.1 and cudnn 7.0.5 on ubuntu x64.

**Comment by nluehr** [comment]
There are multiple fixes available for this issue. Eigen sources after hash 034b6c3 remove the direct reference to math_functions.hpp. So bumping the Eigen version is one fix. Also, TF PR #15739 (which has now been merged into master) works around the problem by adding the cuda/inlcude/crt sub-directory to INCLUDES.

**Comment by EKami** [comment]
The `v1.5.0-rc0` release doesn't seems to include a fix before doing the release. I hope they'll add a fix for the next release candidate.

**Comment by reedwm** [comment]
/CC @gunan, is it possible for the fix to be included in 1.5?

**Comment by gunan** [comment]
This is due to eigen.
Correct fix is here: https://github.com/tensorflow/tensorflow/pull/15796

@av8ramit Once the PR above is merged can we cherrypick it to the release (just the 2nd commit in the PR is sufficient).

**Comment by av8ramit** [comment]
I've merged it. @gunan sure, it will be in rc1

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by keerthanpg** [comment]
So, which version of eigen to use?
